<a href="https://colab.research.google.com/github/MatthewFried/Udemy/blob/master/Day6/Day6_Notes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<br>
<br>
<br>
<br>

#Day 6: Regression Modeling for Categorical  Response Variables
<br>
<br>
<br>

## Regression Modeling for Categorical Response Variables

- __Binary Logistic Regression__: The response variable is a __binary categorical variable__, while the explanatory variables can be either continuous, discrete or binary (e.g., dummy variables created from the values of a categorical variable).


- __Multinomial Logistic Regression (MLR)__: The response variable is a categorical variable having __more than two (2) possible values__, while the explanatory variables can be either continuous, discrete or binary. There are many types of MLR models, and the exact type of MLR model you choose to use will be dependent on whether your response variable is __ordinal__ or __nominal__ in nature.


### Binary Logistic Regression

In binary logistic regression, the response variable is __binary__, i.e., it contains data encoded as either '1' (e.g., 'True') or '0' (e.g., 'False'). In other words, logistic regression is used for __classification__ problems.

Logistic regression finds the best fitting __mathematical model__ (not a simple linear equation) to describe the relationship between the binary response variable and the explanatory variable(s). The values it generates are the coefficients of a formula to predict a __logit transformation__ (aka "__log odds"__) which is the logarithm of the odds $p$ of the probability of the presence of the characteristic of interest. 

$logit(p) = ln(p/(1-p)) = b0 + b1x1 + b2x2 + .. + bnxn$

where $p$ = the probability of the __presence__ of a characteristic,

$(1-p)$ = the probability of the __abscence__ of a characteristic,

$b0$ = is a constant value

$b1, b2, ..,bn$ = the regression coefficients for the explanatory variables $x1, x2, .., xn$

Effective binary logistic regression models contain little to no collinearity amongst the explanatory variables used.

The explanatory variables should be linearly related to the log odds of the probability of the presence of the characteristic of interest (i.e., the value of the response variable)

Logistic regression generally does not work well when applied to very small data sets. For observational studies, a minimum of __500 samples/observations__ is recommended.

Further explanation is available here:

https://www.statisticssolutions.com/what-is-logistic-regression/



### Multinomial Logistic Regression

In multinomial logistic regression (MLR), the response variable is __multiclass__, i.e., it has __more than 2 possible values__. Multinomial regression calculates the probability of any observation being assigned to each of the possible classes. Therefore, the regression model must perform at least $n-1$ calculations, where $n$ is the number of possible classifications for the response variable.

Why $n-1$ calculations? If we have $n$ classes, we can determine the probability of the $n$th class by simply summing the probabilities for each of the other classifications and subtracting that sum from $1$, i.e., 

### $P(Class_n) = 1 - ( P(Class_1) + P(Class_2) + ... + P(Class_{n-1}) ) $

As with binary logistic regression, the coefficients of a MLR model predict the __log odds__ of the probability of any observation belonging to each of the $n$ classes. 

The probability of observing class $k$ out of $n$ total classes is:

# $ P(y_i = k) = \frac { e^{S_k} }  {e^{S_1} + e^{S_1} + ... + e^{S_1} } $

(see https://towardsdatascience.com/understanding-logistic-regression-coefficients-7a719ebebd35)

When performing multiple logistic regression, the algorithm assigns a predicted classification for a given observation based on __which of the possible classifications achieves the largest log odds value__.


### How to interpret logistic regression model coefficients

"A one unit increase in explanatory variable $x$ (with all other explanatory variables held constant) increases the log-odds of response variable $y$ by the amount indicated by the coefficient of $x$."

However, in practice it is very common to simply examine the __directionality__ of the coefficients, i.e., if a coefficient is __negative__, then the larger the value of the explanatory variable, the more it will tend to __decrease__ the magnitude of the response variable. Conversely, if a coefficient is __positive__, then the larger the value of the explanatory variable, the more it will tend to __increase__ the magnitude of the response variable.


### How to convert log odds to a probability

- __Step 1__: Convert the log odds value to odds via __exponentiation__, i.e., the antilog of a log value. Assuming the log odds value we've been given is a __natural log__, we apply an __exponential function__ to convert the log odds value to and odds value.


- __Step 2__: Calculate the probability as follows: 

### $ Prob = \frac {odds} {1 + odds} $

http://www.pmean.com/13/predicted.html


### Advantages of Logistic Regression

- Relatively computationally efficient + easy to implement


- Relatively easy to interpret (though not as easy to interpret as linear regression models)


- Can be effective even if features have not been scaled to similar ranges


- Due to its simplicity, logistic regression is widely used as a baseline when comparing the performance of other more complex classification methods (e.g., decision trees, random forest, KNN, neural networks, etc.)


### Disadvantages of Logistic Regression

- Due to its simplicity, performance can lag behind that of more complex classification methods


- Requires a __linear__ relationship between the explanatory variables and the response variable since it relies on a linear decision space


- Requires that explanatory variables are __very__ independent of one another + that they have a substantive relationship with the response variable.

### Evaluating Logistic Regression Model Performance

We learned about a wide variety of classification model performance metrics, all of which can be used for purposes of evaluating the performance of any type of logistic regression model:

- Confusion Matrices: True Positives (TP), False Positives (FP), True Negatives (TN), False Negatives
 (FN)
 
 
- Accuracy


- Precision 


- Recall / Sensitivity


- Specificity


- F1 Score


- ROC: __ROC (Receiver Operating Characteristic) Curve__: Calculate by plotting the __true positive rate (TPR)__ against the __false positive rate (FPR)__; http://www.saedsayad.com/model_evaluation_c.htm. Plotting TPR vs. FPR for a series of classification models constructed using different thresholds for predicting classifications yields a curve within a two-dimensional plane.


- AUC: In a ROC plot, AUC is determined by calculating the area in the plot that falls __below / to the right of__ the ROC curve.  A random classifier has an area under the curve of 0.5, while AUC for a perfect classifier is equal to 1. The higher the AUC score, the better the performance of a model.  AUC scores from different models can be compared against one another to help us determine which model has the best performance.

## Logistic Regression example with scikit-learn

Here is the 'Titanic' data set which contains passenger survival rate data for the ocean-going vessel of that name that sank in 1912. 

Our goal is to __try to predict whether or not a passenger would have been more likely than not to survive the sinking of the Titanic__ based on the data contained in the data set (i.e., build a __predictive model__).

- First, we will fit a logistic regression model on a __training__ data set using the data set's __PClass__, __Age__, and __Sex__ attributes.


- Then we evaluate the model using a __testing__ data set that is exclusive of the data contained in the __training__ data set.


In [ ]:
# load the LogisticRegression() function from sklearn's 'linear_model' sub-library
from sklearn.linear_model import LogisticRegression

# load the training + testing data sets for the Titanic data
train = pd.read_csv('https://raw.githubusercontent.com/wesm/pydata-book/2nd-edition/datasets/titanic/train.csv')
test = pd.read_csv('https://raw.githubusercontent.com/wesm/pydata-book/2nd-edition/datasets/titanic/test.csv')

# display the first four rows of the data set
train[:4]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S

In [ ]:
# lets check the first four rows of the test data
# Note the lack of a "Survived" indicator !!!
test[:4]

PassengerId  Pclass                              Name     Sex   Age  SibSp  \
0          892       3                  Kelly, Mr. James    male  34.5      0   
1          893       3  Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
2          894       2         Myles, Mr. Thomas Francis    male  62.0      0   
3          895       3                  Wirz, Mr. Albert    male  27.0      0   

   Parch  Ticket    Fare Cabin Embarked  IsFemale  
0      0  330911  7.8292   NaN        Q         0  
1      0  363272  7.0000   NaN        S         1  
2      0  240276  9.6875   NaN        Q         0  
3      0  315154  8.6625   NaN        S         0

In [ ]:
# how many records in the training data set?
train.shape

(891, 12)

In [ ]:
# how many people survived in the training set?
train.Survived.values.sum()

342

In [ ]:
# what percentage of the training set survived?
train.Survived.values.sum() / train.shape[0]

0.3838383838383838

__NOTE__: Since we know that 38.3% of the people in the training set survived, we could achieve a training model accuracy of (1 - .383) = 61.7% by simply predicting "Did not survive" for each passenger. This metric is referred to as the __null error rate__.  When evaluating the performance of a binary logistic regression model, always check to see whether the accuracy you are attaining exceeds the __null error rate__. If not, your model is unlikely to be of any value.


Missing data will generally prevent the fitting of a model via either __scikit-learn__ or __statsmodels__, so we must first check both the training and testing data for missing data values:

In [ ]:
# check the training data for null values
train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Within the training data, the __Age__, __Cabin__ and __Embarked__ attributes all have missing data.

In [ ]:
# check the test data for null values
test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

Within the testing data, the __Age__, __Fare__ and __Cabin__  attributes all have missing data.

So if we want to use __Age__ as part of our predictive model, we need to somehow fill the missing __Age__ values. In this instance, the author of the PfDA text chooses to use the relatively simple process of filling the missing values with the __median__ Age value found within the __training__ data set:

In [ ]:
# find the median 'Age' value within the training data set
impute_value = train['Age'].median()

# now fill the missing 'Age' values in both the training and testing data sets
train['Age'] = train['Age'].fillna(impute_value)
test['Age'] = test['Age'].fillna(impute_value)

Next, __we create a dummy indicator for the 'Sex' categorical variable__: the new dummy variable 'IsFemale' contains a '1' if the 'Sex' value for a passenger is 'Female' and a '0' otherwise:

In [ ]:
# create a dummy variable for the 'Sex' attribute in both the training and
# testing data sets
train['IsFemale'] = (train['Sex'] == 'female').astype(int)
test['IsFemale'] = (test['Sex'] == 'female').astype(int)

Now we are ready to define our predictive model using the attributes we want to use as explanatory variables:

In [ ]:
# define a vector containing the names of the attributes to use
predictors = ['Pclass', 'IsFemale', 'Age']

# create a subset of the training data using ONLY the selected explanatory variables
X_train = train[predictors].values

# create a subset of the testing data using ONLY the selected explanatory variables
X_test = test[predictors].values

# isolate the response indicator for the training data
y_train = train['Survived'].values

# sanity check on training data
X_train[:5]

array([[ 3.,  0., 22.],
       [ 1.,  1., 38.],
       [ 3.,  1., 26.],
       [ 1.,  1., 35.],
       [ 3.,  0., 35.]])

In [ ]:
# sanity check on response indicator
y_train[:5]

array([0, 1, 1, 1, 0], dtype=int64)

In [ ]:
# We're using the LogisticRegression() method for this model
model = LogisticRegression()

# fit the model: X_train contains our explanatory variables while 
# y_train contains the response variable
model.fit(X_train, y_train)

C:\Users\James T\Anaconda3\envs\DAV5400\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [ ]:
# calculate the accuracy of the model relative to the training data set
model.score(X_train, y_train)

0.7946127946127947

Recall from above we calculated the __null error rate__ for our data to be 61.7%. The model we've generated has an accuracy score of 79.46%. As such, our model appears to be useful.

__What can we learn by examining the regression model coefficients for the explanatory variables?__

While the actual numeric values of the coefficients of a regression model are often very difficult to interpret, we can use the fact that a coefficient is either positive or negative to determine what effect of an __increase__ in the value of a given variable will have on the predicted value of the response variable. 

In this binary logistic regression example, the value of the response variable is the "log odds" of the binary outcome being either a '0' or '1' ('0' meaning the passenger was more likely to perish while '1' indicates the passenger was more likely to survive).

In [ ]:
# examine the model coefficients for the explanatory variables
print(predictors)
model.coef_

['Pclass', 'IsFemale', 'Age']


array([[-1.07373582,  2.52093733, -0.02864864]])

From the above we see that:

- __Passenger Class__: An increase in the value of 'Passenger_Class' is associated with a __decreased__ likelihood of survival, i.e., passengers in lower classes of service were more likely to perish than were passengers in relatively higher classes of service.


- __IsFemale__: Being female increased the likelihood of survival


- __Age__: The older the passenger, the less likely they were to survive.

In [ ]:
# generate predictions for the test data using our new model
y_predict = model.predict(X_test)
y_predict[:10]

array([0, 0, 0, 0, 1, 0, 1, 0, 1, 0], dtype=int64)

The 'y_predict' array contains predictions that answer the question: "__Was a given passenger more likely than not to have survived the sinking of the Titanic?__"

If we had the actual __survived/perished__ indicators for the test data, we could now check the performance of the model via a variety of error metrics (e.g., accuracy, specificity, precision, recall, AUC, etc.).

### Case Study: Logistic Regression with scikit-learn

[Link](https://nbviewer.jupyter.org/gist/justmarkham/6d5c061ca5aee67c4316471f8c2ae976)

# **Day6 Assignment**

##Logistic Regression 

Banks are highly regulated. They must adhere to strict lending guidelines on several fronts. The must avoid discrimination when  approving or rejecting loan applications. 

Many years ago, banks engaged in the practice of “redlining” when approving  loans for home mortgages.  They would reject loan applications for homes located  in certain areas that were deemed by banks to be overly prone to loan defaults. When such practices were deemed to  be discriminatory by bank regulators, banks adjusted their mortgage lending practices in an attempt to make them more  objective. However, regulators remain concerned that discrimination can exist even when what appear to be purely  objective criteria are used for purposes of deciding whether or not a customer is approved for a loan. 

Develop a model that can predict whether or not a given mortgage  loan application is likely to be approved or denied.

The data set you will be using is sourced from the Federal Reserve Bank of Boston. [Here](https://vincentarelbundock.github.io/Rdatasets/doc/Ecdat/Hdma.html) is a brief overview of the data.

It has 2,381 observations of 1 response/dependent variable (which indicates whether or  not a mortgage application was denied) and 12 explanatory/independent variables. 

Construct and compare/contrast a series of binary logistic regression  models (after completing the necessary EDA and data prep work) that predict whether or not a given  mortgage application is likely to be denied. The response variable is the data set’s  “DENY” attribute, which indicates whether or not a mortgage application was denied. 

###Process

1. Load the [data](https://raw.githubusercontent.com/MatthewFried/Udemy/master/Day6/Day6%20Data%20-%20Federal%20Reserve.csv) into a pandas DataFrame and perform the EDA. Do the data preparation work. Make  sure to Describe and show the steps you have taken to address the data integrity and usability issues you identified in your EDA, including any feature engineering techniques you have  applied to the data set. 

2. Do feature selection and/or dimensionality reduction techniques to identify  explanatory variables for inclusion within your models. You should apply domain knowledge, use forward or backward selection, or use a different feature  selection method (e.g., decision trees, etc.). 

3. After splitting the data into training and testing subsets, use the training subset to construct at least  three different binomial logistic regression models using different combinations of explanatory  variables (or the same variables if they have been transformed via different transformation methods). 

4. Decide how you will select the “best” regression model from those  you have constructed. For example, are you willing to select a model with slightly lower performance if  it is easier to interpret or less complicated to implement? What metrics will you use to  compare/contrast your models? Evaluate the performance of your models via cross validation using  the training data set. Then apply your preferred model to the testing subset and assess how well it  performs on that previously unseen data. 


##**Review of Assignment**

The __deny__ responsible variable is an __imbalanced class__: more than 88% of its values are '0' while less than 12% are '1'. The __null error rate__ for the response variable is 0.882, which means that we could achieve 88.2% accuracy by simply predicting that a loan is not denied for every observation within the data set. Obviously, such a model wouldn't be very useful for purposes of deciding whether or not a mortgage application should be denied.

We learned that reliance on __accuracy__ for purposes of comparing similar models __is not appropriate when the response variable is an imbalanced class__. 

Which metrics should we rely on?

- __precision__: TP / (TP + FP) If we are trying to __minimize the number of false positives (FP)__, we should use __precision__ as one of our primary model performance metrics.


- __recall__: TP / (TP + FN) If we are trying to __maximize the number of true positives (TP)__ or __minimize the number of false negatives__, we should use __recall__ as one of our primary model performance metrics. Why? Maximization of TP necessarily minimizes instances of FP.


- __F1 Score__: 2 * (precision * recall) / (precision + recall) is the __weighted average of precision and recall__. Therefore, F1 score __is a measure of how well a model handles both false positives and false negatives__. Remember: models with relatively larger F1 scores are preferable to models having relatively smaller F1 scores.


#### How else might we handle imbalanced classes?

- Synthetic Minority Oversampling Technique ("SMOTE"): The concept of SMOTE was explained in the Module 4 Assigned readings (see __Machine Learning Pocket References, Chapter 9__). SMOTE works by __synthesizing__ new examples from the minority class. The process works as follows:

- A random example from the minority class is chosen. 


- k of the nearest neighbors for that example are found (typically k=5). 


- An additional randomly selected neighbor is chosen and a synthetic example is created at a randomly selected point between the two examples in feature space.


This process is repeated as many times as needed to balance out the classifications for the imbalanced variable

The approach is effective because new synthetic examples from the minority class are created that are __plausible__, meaning their features are relatively to those of existing examples from the minority class.

When finished, we then have a balanced class. So if our variable is a binary categorical feature, the __null error rate__ for the data that includes the synthesized observations will be __.50__. With a .50 null error rate, an accuracy metric can be a very effective tool for comparing models.

For more details + examples see this [link](https://machinelearningmastery.com/smote-oversampling-for-imbalanced-classification/)

#### So should we always use SMOTE if we have an imbalanced response variable?

Not necessarily. It can be effective in some instances but is not guaranteed to improve model performance. As always, we need to test it empirically to judge its efficacy relative to a specific data set.
